### ISSUE TITLE
Switching server database at runtime exposes data to wrong user

### ISSUE THREAD

**Issue by VagueGit** [Mimetis/Dotmim.Sync#598]
Our customers are companies and each company has their own database on our server. We have rolled out a version of our app that uses DMS to sync SQLite on the client to their database via https.

The client passes a token in the header `httpClient.DefaultRequestHeaders.Authorization = new AuthenticationHeaderValue(...`

The sync client is configured
`WebClientOrchestrator webClientOrchestrator = new WebClientOrchestrator(serviceUri, client: httpClient, maxDownladingDegreeOfParallelism: 4);`

On the server the `SyncController `uses the token to lookup which database to use for that user.
```
string databaseName = DatabaseHelper.GetDatabaseName(token);
SqlConnectionStringBuilder builder = new SqlConnectionStringBuilder(orchestrator.Provider.ConnectionString);
builder.InitialCatalog = databaseName;
orchestrator.Provider.ConnectionString = builder.ConnectionString;
```

We've rolled out this version of our app to less than 100 customers as beta testers. We've had one report of a customer seeing some data that was not theirs, mixed in with theirs. We immediately shut down the beta test.

Where is the flaw in our strategy for switching databases at runtime. Is there recommended practice?

**Comment by VagueGit** [comment]
Thank you for taking the time for such a comprehensive explanation. To answer your questions in order:

1. We don't use filters.
2. As 1.
3. Each customer starts with an empty SQLite database on the client (and an empty SQL Server database on the server).
4. `Reinitialize` is not called explicitly. The empty client database is initialised by DMS when `agent.SyncAsync` is called using default scope. Default scope is all tables.
5. We use an existing, empty client database.

All customers start with a copy of exactly the same empty SQLite database and a copy of the same empty SQL Server database. We use `Dotmim.Sync.SqlServer.ChangeTracking`

DMS then initialises the empty SQLite database with agent.SyncAsync using default scope.
The app then populates the SQLite database with existing customer data on their machine.
DMS then syncs the populated SQLite database to the server.

When we started the beta, multiple customers synced their data to the server simultaneously. It was after the initial sync that a customer reported seeing another customer's data. The data was not scrambled; it was intact. Complete customer orders for one customer were visible to another customer. Obviously this is a serious breach, so we shut down the beta immediately.

We are fortunate this did not happen in production. We will be able to test further if you can offer guidance. Please let me know if you require further information..

**Comment by Mimetis** [comment]
So, If i'm right, each client has its own server database.
It's a 1-1 relationship, right ?
I mean, each server database is only used by one single customer ?

In a nutshell, the server database is basically a copy of the customer sqlite database ?

If i'm right, if a customer has some rows from another customer, it's because the sync from server was not targetting the good server database.

I guess, you have a problem with the connection string allocation.
Are you sure you did not share the same connection with 2 customers ?

**Comment by VagueGit** [comment]
Again, in order:

1. Each customer has it's own server database. Each customer may have multiple computers connecting to that server database, using the same login.
2. See 1.
3. Each server database is only used by one customer.
4. The SQLite database is a copy of the server database. We generate the SQLite database from a script which reads the attributes of the SQL Server database and generates a SQLite database. This allows us to apply all necessary constraints to the SQLite database. We do this once. The app downloads a copy of the SQLite database from our CDN.
5. The same server database cannot be shared among customers as the database name is retrieved by the `SyncController` as shown earlier in this thread, using the request authentication header. The authentication header is generated from the customer login. The customer login is email address and password. The user email address is validated before the account and server database are created.

**Comment by Mimetis** [comment]
Well, I don't see where the collision is occurring to be honest....
Are you able to reproduce it ?

**Comment by VagueGit** [comment]
I am not able to reproduce it in-house. We've been testing in-house for some time without this issue arising. It occurred when we released to ~100 beta testers.

We can only test in-house until this issue is resolved. We can't legally risk exposing our customers' data to other customers.

**Comment by VagueGit** [comment]
DMS source is too sophisticated/complex for me to get my head around well enough. Does DMS now guarantee a PER REQUEST instance as @gentledepp is using in an earlier version of DMS in production?. With my limited understanding it appears if that were so then leakage of data from one customer to another would not be possible.

**Comment by Mimetis** [comment]
I have changed the way the server is handling information for each client.

I've removed all the **server cache usages**, and replace them with the **client session**.

And for your particular need, I've changed a little bit the `WebServerOrchestrator` to let you create one dynamically in the controller.

So far, if you want to test it, just clone the last commit from the master branch and test it.

It's not mandatory to migrate your clients, since the changes are only occurring on the server side (if your clients are already on version `v0.8.0`

Since we are using now the session, you need to add the session service and middleware.
See here for more info : https://docs.microsoft.com/en-us/aspnet/core/fundamentals/app-state?view=aspnetcore-5.0#configure-session-state.

Here is what you should have now:

**Startup.cs:**

``` cs
public void ConfigureServices(IServiceCollection services)
{
    services.AddControllers();

    services.AddDistributedMemoryCache();
    services.AddSession(options =>
    {
        options.IdleTimeout = TimeSpan.FromMinutes(30);
    });
}

```

Be careful, order is important here : `UseSession `is **AFTER** `UseRouting `and `UseAuthorization `and **BEFORE** `UseEndpoints`

``` cs
// This method gets called by the runtime. Use this method to configure the HTTP request pipeline.
public void Configure(IApplicationBuilder app, IWebHostEnvironment env)
{
    if (env.IsDevelopment())
    {
        app.UseDeveloperExceptionPage();
    }
    app.UseHttpsRedirection();
    app.UseRouting();
    app.UseAuthorization();
    app.UseSession();
    app.UseEndpoints(endpoints =>
    {
        endpoints.MapControllers();
    });
}
```

And for your particular use case, where you need to create a new Orchestrator for each request, with a different connection string:

**SyncController.cs**:

``` cs
public SyncController(WebServerManager webServerManager, IConfiguration configuration)
{
    this.webServerManager = webServerManager;
    this.Configuration = configuration;
}

[HttpPost]
public async Task Post()
{
    var connectionString = Configuration.GetSection("ConnectionStrings")["SqlConnection"];

    var tables = new string[] {"ProductCategory", "ProductModel", "Product",
    "Address", "Customer", "CustomerAddress", "SalesOrderHeader", "SalesOrderDetail" };
    var provider = new SqlSyncChangeTrackingProvider(connectionString);

    var orchestrator = new WebServerOrchestrator(provider, tables);

    await orchestrator.HandleRequestAsync(this.HttpContext);
}

```
Let me know if it's resolving your issue.
It will be quite complicated for you, I know, since you have to redeploy and retests with 100 customers, but it's complicated for me as well since I'm not able to reproduce your bug, and once again, you don't share a lot of your code to help me figure out what can be the issue in your code...

With this fix, I'm almost 100% sure that nothing is shared between sessions.

**Comment by VagueGit** [comment]
Thank you @Mimetis . When do you expect this version to appear in the DMS Nuget pckage?

**Comment by Mimetis** [comment]
I'm trying to fix some weird bugs, and be sure all the tests are OK
Once it's done, I will probably make a new version in nuget
I don't know, maybe next week

**Comment by Mimetis** [comment]
You can test the beta published version `v0.8.1-beta-0687`
Please take a look at https://github.com/Mimetis/Dotmim.Sync/discussions/602 for info / upgrade

Let me know if it's resolving your issue

**Comment by VagueGit** [comment]
We've released this beta to 50+ beta testers and the issue does appear to be resolved. Thank you.

We'd like to incorporate this package into one of our commercial releases. When do you expect this fix to be out of beta?
